Libraries/LibWeb/Editing/Commands.cpp

```diff
@@ -384,9 +368,347 @@ bool command_delete_action(DOM::Document& document, String const&)
+        if (end_of_line && is_heading(container_element.local_name()))
+            return document.default_single_line_container_name();
+
+        // 19. Otherwise, if the local name of container is "dt" and end of line is true, let new container name be "dd".
+        if (container_element.local_name() == HTML::TagNames::dt && end_of_line)
+            return HTML::TagNames::dd;
+
+        // 20. Otherwise, if the local name of container is "dd" and end of line is true, let new container name be "dt".
+        if (container_element.local_name() == HTML::TagNames::dd && end_of_line)
+            return HTML::TagNames::dt;
+
+        // 21. Otherwise, let new container name be the local name of container.
+        return container_element.local_name();
+    }();
+
+    // 22. Let new container be the result of calling createElement(new container name) on the context object.
+    GC::Ptr<DOM::Element> new_container = MUST(DOM::create_element(document, new_container_name, Namespace::HTML));
+
+    // 23. Copy all attributes of container to new container.
+    container_element.for_each_attribute([&new_container](FlyString const& name, String const& value) {
+        MUST(new_container->set_attribute(name, value));
+    });
+
+    // 24. If new container has an id attribute, unset it.
+    if (new_container->has_attribute(HTML::AttributeNames::id))
+        new_container->remove_attribute(HTML::AttributeNames::id);
+
+    // 25. Insert new container into the parent of container immediately after container.
+    container->parent()->insert_before(*new_container, container->next_sibling());
+
+    // 26. Let contained nodes be all nodes contained in new line range.
+    // FIXME: this is probably wildly inefficient
+    Vector<GC::Ref<DOM::Node>> contained_nodes;
+    auto common_ancestor = new_line_range->common_ancestor_container();
+    common_ancestor->for_each_in_subtree([&](GC::Ref<DOM::Node> child_node) {
+        if (new_line_range->contains_node(child_node))
+            contained_nodes.append(child_node);
+        return TraversalDecision::Continue;
+    });
+
+    // 27. Let frag be the result of calling extractContents() on new line range.
+    auto frag = MUST(new_line_range->extract_contents());
+
+    // 28. Unset the id attribute (if any) of each Element descendant of frag that is not in contained nodes.
+    frag->for_each_in_subtree_of_type<DOM::Element>([&contained_nodes](GC::Ref<DOM::Element> descendant) {
+        if (!contained_nodes.contains_slow(descendant))
+            descendant->remove_attribute(HTML::AttributeNames::id);
+        return TraversalDecision::Continue;
+    });
+
+    // 29. Call appendChild(frag) on new container.
+    MUST(new_container->append_child(frag));
+
+    // 30. While container's lastChild is a prohibited paragraph child, set container to its lastChild.
+    while (container->last_child() && is_prohibited_paragraph_child(*container->last_child()))
+        container = container->last_child();
```

Comment:
Could `last_child()` return null here?

Reply:
Well, no? That's why there's a null check in the `while`?